feature/experimental/gribi/otg_tests/programming_with_reload/programming_with_reload_test.go

```diff
@@ -0,0 +1,511 @@
+		inPkts := float32(gnmi.Get(t, ate.OTG(), gnmi.OTG().Flow(flow.Name()).Counters().InPkts().State()))
+		if outPkts == 0 {
+			t.Fatalf("OutPkts for flow %s is 0, want > 0", flow)
+		}
+		if got := ((outPkts - inPkts) * 100) / outPkts; got < 100 {
+			t.Fatalf("LossPct for flow %s: got %v, want 100", flow.Name(), got)
+		}
+	}
+}
+
+// setDUTInterfaceState sets the admin state on the dut interface
+func (args *testArgs) setDUTInterfaceWithState(t testing.TB, p *ondatra.Port, state bool) {
+	dc := gnmi.OC()
+	i := &oc.Interface{}
+	i.Enabled = ygot.Bool(state)
+	i.Type = ethernetCsmacd
+	i.Name = ygot.String(p.Name())
+	// Following reload try gNMI call twice, Unavaiable error code is returned for the 1st time
+	for {
+		if errMsg := testt.CaptureFatal(t, func(t testing.TB) {
+			gnmi.Update(t, args.dut, dc.Interface(p.Name()).Config(), i)
+		}); errMsg != nil {
+			t.Logf("Got testt.CaptureFatal errMsg: %s, retrying ...", *errMsg)
+		} else {
+			break
+		}
+	}
+}
+
+// addStaticRoute configures static route needed for decap path
+func addStaticRoute(t *testing.T, dut *ondatra.DUTDevice, ip string) {
+	d := gnmi.OC()
+	s := &oc.Root{}
+	gnmi.Update(t, dut, d.NetworkInstance(deviations.DefaultNetworkInstance(dut)).Config(), &oc.NetworkInstance{
+		Name: ygot.String("DEFAULT"),
+	})
+	static := s.GetOrCreateNetworkInstance(deviations.DefaultNetworkInstance(dut)).GetOrCreateProtocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(dut))
+	ipv4Nh := static.GetOrCreateStatic(innerDstIP1 + "/" + mask).GetOrCreateNextHop("0")
+	ipv4Nh.NextHop, _ = ipv4Nh.To_NetworkInstance_Protocol_Static_NextHop_NextHop_Union(ip)
+	gnmi.Update(t, dut, d.NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(dut)).Config(), static)
+}
+
+// Reload chassis
+func (args *testArgs) reloadChassis(t *testing.T) {
+
+	gnoiClient, err := args.dut.RawAPIs().BindingDUT().DialGNOI(context.Background())
+	if err != nil {
+		t.Fatalf("Error dialing gNOI: %v", err)
+	}
+	bootTimeBeforeReboot := gnmi.Get(t, args.dut, gnmi.OC().System().BootTime().State())
+	t.Logf("DUT boot time before reboot: %v", bootTimeBeforeReboot)
+
+	rebootRequest := &spb.RebootRequest{
+		Method:  spb.RebootMethod_COLD,
+		Delay:   0,
+		Message: "Reboot chassis with delay",
+		Force:   true,
+	}
+
+	t.Logf("Send reboot request: %v", rebootRequest)
+	rebootResponse, err := gnoiClient.System().Reboot(context.Background(), rebootRequest)
+	t.Logf("Got reboot response: %v, err: %v", rebootResponse, err)
+	if err != nil {
+		t.Fatalf("Failed to reboot chassis with unexpected err: %v", err)
+	}
+
+	// adding sleep time for reload to begin
+	time.Sleep(1 * time.Minute)
+
+	// establish new gribi programming following reload
+	err_msg := args.gribi_reconnect(t)
+	if err_msg != nil {
+		t.Fatalf("Failure: %d", err_msg)
+	}
```

Comment:
can we also add check to make sure that gnmi server is up and reconnected . 

Also , need to add check for all interface that are part of up test are up .  we can loop gnmi.await() some thing like 
''''
	for p, _ := range dutPorts {
		p1 := args.dut.Port(t, p)
		gnmi.Await(t, args.dut, gnmi.OC().Interface(p1.Name()).OperStatus().State(), 2*time.Minute, oc.Interface_OperStatus_UP)
	}
'''

Reply:
We are not going to check interface or hw status following reload as it is not part of this tc as indicated by google in previous comments

Reply:
@manan-patel  ,  Not sure which comment you are talking about it . But we need to make sure that gNMI connection exceptions are handled after reload .  For example L280 gnmi set will fail if gnMI server was not up . Alternatively we can add gNMI connection exception handling in L280 with retry option. 

If we don't check and wait for interfaces up then traffic  check in L288 can lead to false failure. 
@xw-g  could you please help with your inputs.

Reply:
I removed the post checks as per comment on Nov 22, 2023 by @LimeHat, as per suggested I can uplift the tc accordingly

Reply:
Hi @manan-patel @bkreddy143 

AFAIR there was never a check on the interface status.
- There was a check that verifies the length of the component list with any oper-status (that wasn't particularly useful here)
- Also, there was a check that used to verify gnmi status before starting gRIBI reconnect attempts 

Indeed I commented on both points.

@bkreddy143 request is slightly different:

- to make sure that interfaces are up before you start traffic validation (which makes sense. alternatively, you need to remove the traffic validation part). This can be done after gribi reconnects, and only for interfaces in binding 
- also, if there are any gnmi requests in the subsequent steps (following the gribi reconnect), we'll need to wait for gnmi server to be up accordingly

I.e. my comments were mainly about the fact that we should start gribi reconnection attempts pretty much immediately post reboot (they are not blocked by gnmi or anything else), but @bkreddy143 correctly notes among the same lines that all three parts of this equation (gribi, gnmi, ports oper status) are independent (in a general case), and you need to monitor status for each before you attempt to execute it.